packages/lesswrong/server/search/elastic/searchQuery.ts

```diff
@@ -0,0 +1,27 @@
+import { z } from "zod";
+
+const querySchema = z.object({
+  indexName: z.string(),
+  params: z.object({
+    query: z.string(),
+    highlightPreTag: z.optional(z.string()),
+    highlightPostTag: z.optional(z.string()),
+    hitsPerPage: z.optional(z.number().int().nonnegative()),
+    page: z.optional(z.number().int().nonnegative()),
+    facetFilters: z.optional(z.array(z.array(z.string()))),
+    numericFilters: z.optional(z.array(z.string())),
+  }),
+});
+
+export type SearchQuery = z.infer<typeof querySchema>;
+
+export const isValidSearchQuery = (value: unknown): value is SearchQuery => {
+  try {
+    querySchema.parse(value);
+    return true;
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error("Invalid search query:", e.message, JSON.stringify(value, null, 2));
+    return false;
+  }
```

Comment:
I think this could also be:
```
const result = querySchema.safeParse(value);
if (!result.success) {
  console.error("Invalid search query:", result.error.message, JSON.stringify(value, null, 2));
  return false;
}
return true;
```

which is a bit nicer both because it doesn't involve try/catch and also `result.error` is typed as a `ZodError`.

Reply:
(After having looked at the `zod` documentation the last time this came up, I do think it's probably better than `io-ts` given that we aren't really going to be using `fp-ts` in any case - `zod` seems to handle optional fields better, among other things.)